### ISSUE TITLE
"/"-prefixed part files fails to compile but is fine in analyzer

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#52964]
```dart
// lib/fruits.dart
part '/src/fruits/apple.dart';
```

```dart
// lib/src/fruits/apple.dart
part of '../fruits.dart';

class Apple {}
```

```dart
// lib/main.dart
import '/fruits.dart';

void main() {
  Apple();
}
```

This program appears fine in the analyzer, but fails to compile (in the CFE?) citing that "Apple" cannot be found/resolved. It appears the CFE (?) doesn't resolve part files the same as the analyzer when using `/`-prefixed URIs.

**Comment by johnniwinther** [comment]
@lrhn Doesn't `/` in the beginning of the path mean that the path is absolute? In which case the lookup _shouldn't_ resolve to a Uri relative to the current Uri?

**Comment by lrhn** [comment]
(The `part of` needs another `..`, so `part of '../../fruits.dart';`, to get out of both the `fruits` and `src` directories. That's not the problem, but it is one.)

A URI reference of `/fruits.dart` is an absolute *path*, but not an absolute URI. It has no authority and no scheme, so if resolved against another URI, it should still inherit those from the base.

That is, when resolved against a full base URI, for example `http://example.com/banana/boat.dart`, the relative URI reference `/fruits.dart` would/should give `http://example.com/fruits.dart`.

The rules for resolving against `package:` URIs is not as formally specified as I'd like, but I've implemented the canonical implementation in the Dart `Uri` class. The implementation if the `Uri` class *should* treat package URIs specially, and treat the package name as not part of the path when doing resolution. (But not when asking for `uri.path`, because that would be too breaking. I'd love to move the package name to be accessed through `host` instead, but too many things depend on the current behavior).

So with a base URI of `package:my_package/banana/boat.dart`, the `/fruits.dart` path *should* give `package:my_package/fruits.dart`.

That's if things work correctly, and the code actually uses the `Uri` class to do the resolution. (And the special behavior only works for valid `package:` URIs, a `package:` URI cannot have an authority part, the one with host-name, port etc., and cannot have an actual URI path which starts with `/`. If you create one, like `Uri.parse('package://example.com/banana/boat.dart')`, it will not be considered a valid package URI, and will not be subject to the special logic. I doubt that's what's happening here, but I'd have to see the URI being resolved against to be certain.)

The `Uri` class *seems* to work when I check it.

So is the front-end not using `Uri` to resolve URIs against the library's `package:` URI, *or* are they not using a `package:` URI to begin with, because they already resolved it to a file URI? (If you resolve `/fruits.dart` against a `file:///some/long/file.dart` URI, you get `file:///fruits.dart` which won't be correct.)


If I run `dart lib/main.dart` or `dart run lib/main.dart`, I get the error. The *location* of that error is:
```
lib/fruits.dart:2:6: Error: Error when reading '/src/fruits/apple.dart': No such file or directory
part '/src/fruits/apple.dart';
     ^
```
aka `lib/fruits.dart`, which looks like a _relative path_, not a package URI. That's suspicious.
(And it's defintely the front-end, using `dart compile js` doesn't change anything, so it's not, e.g., VM only.)(


The front end _should_, like the analyzer seems to do, and like the front-end itself does in *some* cases, see that `lib/main.dart` is actually a packge file (it's in the `lib` directory of a package of the current package configuration), and the file should therefore be identified by the URI `package:fruits/main.dart`.

It should not use the `lib/main.dart` file path from the command line when running files inside `lib/`.
We used to do that, and it the stream of issues asking "why doesn't this work correctly" was a clear sign that it didn't work. We changed the parser so that if your entry script file is inside `lib/`, we convert its URI to a `package:` URI before loading and parsing it. (But after finding the package resolution config file relative to the original file location.)

It's actually _worse_ than that. The front end is not just using the wrong URI by accident, it's deliberately throwing the correct URI away, even if it has it.

Even if I *specify* the file using a `package:` URI, by first selecting the package resolution,
*it still doesn't work*:
```
$ dart --packages=.dart_tool/package_config.json package:fruits/main.dart
lib/fruits.dart:2:6: Error: Error when reading '/src/fruits/apple.dart': No such file or directory
part '/src/fruits/apple.dart';
     ^
...
```
It *finds* the `lib/main.dart` file, specified with a `package:fruits/main.dart` URI, then fails to resolve `/src/fruits/apple.dart` correctly against that URI (because if it did, it _would_ find `lib/src/fruits/apple.dart`, which exists).

My bet is on not using `Uri` to do the resolution, and doing some shorthand for replacing the entire path when the `part` file URI starts with `/`. Or assuming that a "relative" URI (one with no scheme) can always be resolved realtive to the `file:` URI of the current library, instead of its canonical `package:` URI, just to save some steps. 
Or something.

In any case, it's wrong, and should be fixed.

Also see: https://github.com/dart-lang/sdk/issues/44151
I think we still have unified our treatment of input script paths.
